### PR TITLE
Implement WorkbenchHelper and Nessus parser

### PR DESCRIPTION
**Pull request by alanfung**
* Support for assets and vulnerability aggregations.
* WorkbenchHelper example in `examples/workbench.py`.

Issue: TAPI-272

### CODE REVIEW

File: tenable_io/api/workbenches.py [hunk 1]
```diff
@@ -1,11 +1,30 @@
...
+from six.moves import xrange
 from tenable_io.api.base import BaseApi
 from tenable_io.api.models import AssetList, AssetInfo, VulnerabilityList, \
     VulnerabilityOutputList
 
 
-class WorkbenchApi(BaseApi):
+class WorkbenchesApi(BaseApi):
+
+    CHAPTER_DIFF = u'diff'
```
Comment (tenableAndrew) [inline tenable_io/api/workbenches.py]: Please remove diff as a chapter option, support for this is going away shortly.

File: tenable_io/api/workbenches.py [hunk 2]
```diff
@@ -67,6 +86,97 @@ def asset_vulnerabilities(self, asset_id, date_range=None, filters=None, filter_
...
+            None.
+        :param asset_id: Restrict the export data to only findings the asset with this id. Note that this id is a UUID,
+            default to None.
+        :raise TenableIOApiException:  When API error is encountered.
+        :return: The file ID.
+        """
+        assert format in [WorkbenchesApi.FORMAT_CSV, WorkbenchesApi.FORMAT_HTML, WorkbenchesApi.FORMAT_NESSUS,
+                          WorkbenchesApi.FORMAT_PDF, ], u'Valid file format.'
+        assert report in [WorkbenchesApi.REPORT_VULNERABILITIES, ], u'Valid report type.'
+        assert chapter in [WorkbenchesApi.CHAPTER_DIFF, WorkbenchesApi.CHAPTER_EXEC_SUMMARY,
```
Comment (tenableAndrew) [inline tenable_io/api/workbenches.py]: Remove CHAPTER_DIFF here too.

File: tenable_io/helpers/workbench.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+        :return: A list of :class:`Vulnerability`.
+        """
+        if asset_id is not None:
+            vulnerabilities = self._client.workbenches_api.asset_vulnerabilities(
+                asset_id, date_range=date_range).vulnerabilities
+        else:
+            vulnerabilities = self._client.workbenches_api.vulnerabilities(date_range=date_range).vulnerabilities
+        return vulnerabilities
+
+    def vulnerabilities_parse(self, date_range=1, asset_id=None, page_size=DEFAULT_PAGE_SIZE):
```
Comment (tenableAndrew) [inline tenable_io/helpers/workbench.py]: I'm seeing some parsing errors throw here for one of my accounts. 

`xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 87553, column 108`

File: tenable_io/api/workbenches.py [hunk 4]
```diff
@@ -67,6 +86,97 @@ def asset_vulnerabilities(self, asset_id, date_range=None, filters=None, filter_
...
+        """Download a file that has been prepared for export.
+
+        :param file_id: The unique identifier of the workbench report being downloaded.
+        :param stream: Default to True. If False, the response content will be immediately downloaded.
+        :param chunk_size: If Stream=False, data is returned as a single chunk.
+            If Stream=True, it's the number of bytes it should read into memory.
+        :raise TenableIOApiException:  When API error is encountered.
+        :return: The content iterator for the file.
+        """
+        response = self._client.get('workbenches/export/%(file_id)s/download',
```
Comment (tenableAndrew) [inline tenable_io/api/workbenches.py]: Should we be checking that the status of the file_id == STATUS_EXPORT_READY before attempting to download?

File: tenable_io/api/workbenches.py [hunk 5]
```diff
@@ -1,11 +1,30 @@
...
+    FILTER_TYPE_OR = u'or'
+
+    FORMAT_CSV = u'csv'
+    FORMAT_HTML = u'html'
+    FORMAT_NESSUS = u'nessus'
+    FORMAT_PDF = u'pdf'
+
+    REPORT_VULNERABILITIES = u'vulnerabilities'
+
+    STATUS_EXPORT_READY = u'ready'
```
Comment (tenableAndrew) [inline tenable_io/api/workbenches.py]: Consider adding the other likely statuses? 

STATUS_EXPORT_LOADING = u'loading'
STATUS_EXPORT_MISSING = u'File does not exist'

File: tests/integration/api/test_workbenches.py [hunk 6]
```diff
@@ -38,3 +41,29 @@ def test_vulnerability_output(self, client):
...
+            WorkbenchesApi.CHAPTER_VULN_BY_ASSET,
+        )
+        assert file_id, u'The `export_request` method returns a valid file ID.'
+
+        export_status = self.wait_until(lambda: client.workbenches_api.export_status(file_id),
+                                        lambda status: status == WorkbenchesApi.STATUS_EXPORT_READY)
+
+        assert export_status == WorkbenchesApi.STATUS_EXPORT_READY, u'Workbench export is ready.'
+
+        iter_content = client.workbenches_api.export_download(file_id, False, None)
```
Comment (tenableAndrew) [inline tests/integration/api/test_workbenches.py]: I'm getting an error when running this test if I include the None as the third param, it passes if I remove this param.

```
============================================================================================= FAILURES =============================================================================================
__________________________________________________________________________________ TestWorkbenchesApi.test_export __________________________________________________________________________________

self = <tests.integration.api.test_workbenches.TestWorkbenchesApi object at 0x1059dfd68>, app = <tests.conftest.App object at 0x1059dfe80>
client = <tenable_io.client.TenableIOClient object at 0x1059fb6a0>

    def test_export(self, app, client):
        file_id = client.workbenches_api.export_request(
            WorkbenchesApi.FORMAT_NESSUS,
            WorkbenchesApi.REPORT_VULNERABILITIES,
            WorkbenchesApi.CHAPTER_VULN_BY_ASSET,
        )
        assert file_id, u'The `export_request` method returns a valid file ID.'
    
        export_status = self.wait_until(lambda: client.workbenches_api.export_status(file_id),
                                        lambda status: status == WorkbenchesApi.STATUS_EXPORT_READY)
    
        assert export_status == WorkbenchesApi.STATUS_EXPORT_READY, u'Workbench export is ready.'
    
        iter_content = client.workbenches_api.export_download(file_id, False, None)
        download_path = app.session_file_output('test_workbench_export')
    
        assert not os.path.isfile(download_path), u'Workbench report does not yet exist.'
    
        with open(download_path, 'wb') as fd:
>           for chunk in iter_content:

tests/integration/api/test_workbenches.py:64: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

string = b'<?xml version="1.0" ?>\n<NessusClientData_v2>\n<Report name="workbenchexport" xmlns:cm="http://www.nessus.org/cm"></ReportHost></Report></NessusClientData_v2>', slice_length = None

    def iter_slices(string, slice_length):
        """Iterate over slices of a string."""
        pos = 0
        while pos < len(string):
>           yield string[pos:pos + slice_length]
E           TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/requests/utils.py:350: TypeError
```

File: tenable_io/helpers/workbench.py [hunk 7]
```diff
@@ -0,0 +1,417 @@
+import tempfile
+
+from tenable_io.api.workbenches import WorkbenchesApi
+from tenable_io.parser.workbenches import WorkbenchParser
+from tenable_io.util import wait_until
+
+
+class WorkbenchHelper(object):
```
Comment (tenableAndrew) [inline tenable_io/helpers/workbench.py]: Should we add a download helper function?

For example, something like this mirrors the scan download function:
```
    def download(self,
                 path,
                 format=WorkbenchesApi.FORMAT_NESSUS,
                 report=WorkbenchesApi.REPORT_VULNERABILITIES,
                 chapter=WorkbenchesApi.CHAPTER_VULN_BY_ASSET,
                 start_date=None,
                 date_range=None,
                 filters=None,
                 filter_search_type=None,
                 minimum_vuln_info=None,
                 plugin_id=None,
                 asset_id=None,
                 file_open_mode='wb'):
        """Download a workbench.

        :param path: The file path to save the export to.
        :param format: The file format, one of FORMAT_CSV, FORMAT_HTML, FORMAT_NESSUS, FORMAT_PDF. Defaults to
            FORMAT_NESSUS.
        :param report: The type of workbench report, one of REPORT_VULNERABILITIES. Defaults to REPORT_VULNERABILITIES.
        :param chapter: Chapter to include, one of CHAPTER_DIFF, CHAPTER_EXEC_SUMMARY, CHAPTER_VULN_BY_ASSET,
            CHAPTER_VULN_BY_PLUGIN, CHAPTER_VULN_HOSTS_SUMMARY. Defaults to CHAPTER_VULN_BY_ASSET.
        :param start_date: The date (in unixtime) at which the exported results should begin to be included, default to
            None which implies today.
        :param date_range: The number of days of data prior to and including start_date that should be returned. default
            to None which implies data for all dates is returned.
        :param filters: An list containing filters to apply to the exported scan report, default to None.
        :param filter_search_type: The type of search to be used, one of FILTER_TYPE_AND, FILTER_TYPE_OR, default to
            None.
        :param minimum_vuln_info: When True, only a minimal subset of scan details will be returned for each result,
            excluding plugin attributes. In this case, only plugin_output and vulnerability_state fields are always
            returned; first_found, last_found and last_fixed are also returned if possible. Default to None.
        :param plugin_id: Restrict the export data to only vulnerabilities found by the plugin with this id, default to
            None.
        :param asset_id: Restrict the export data to only findings the asset with this id. Note that this id is a UUID,
            default to None.
        :param file_open_mode: The open mode to the file output. Default to "wb".
        """

        file_id = self._client.workbenches_api.export_request(format=format,
                                                              report=report,
                                                              chapter=chapter,
                                                              start_date=start_date,
                                                              date_range=date_range,
                                                              filters=filters,
                                                              filter_search_type=filter_search_type,
                                                              minimum_vuln_info=minimum_vuln_info,
                                                              plugin_id=plugin_id,
                                                              asset_id=asset_id)

        wait_until(lambda: self._client.workbenches_api.export_status(file_id) == WorkbenchesApi.STATUS_EXPORT_READY)

        iter_content = self._client.workbenches_api.export_download(file_id)
        with open(path, file_open_mode) as fd:
            for chunk in iter_content:
                fd.write(chunk)

```

File: tenable_io/api/workbenches.py (hunk 4 above)
Comment (tenableAndrew) [inline tenable_io/api/workbenches.py]: We could skip adding a check here if we implement a helper download function.

File: tenable_io/helpers/workbench.py (hunk 3 above)
Comment (alanfung) [inline tenable_io/helpers/workbench.py]: That's not good. Can you send over some sample XML that you might have? I would have to debug this.

File: tenable_io/api/workbenches.py (hunk 4 above)
Comment (alanfung) [inline tenable_io/api/workbenches.py]: Good suggestion, we can add that as a helper level feature. Status checking is not an API-wrapper level concerns.

File: tenable_io/api/workbenches.py (hunk 5 above)
Comment (alanfung) [inline tenable_io/api/workbenches.py]: Do you know if these status values are documented anywhere in the API doc? I've been adding them on an as-needed basis.

File: tenable_io/helpers/workbench.py (hunk 7 above)
Comment (alanfung) [inline tenable_io/helpers/workbench.py]: Yes, let's open an issue and I'll do this on a separate PR?

File: tenable_io/api/workbenches.py (hunk 5 above)
Comment (tenableAndrew) [inline tenable_io/api/workbenches.py]: I don't believe they're documented anywhere. I will file a ticket to add them to our documentation.

File: tenable_io/helpers/workbench.py (hunk 7 above)
Comment (tenableAndrew) [inline tenable_io/helpers/workbench.py]: Sounds good.

File: tenable_io/helpers/workbench.py (hunk 3 above)
Comment (alanfung) [inline tenable_io/helpers/workbench.py]: Seem to be malformed XML when 1. it is an empty set, and 2. when the `<see_also>` contains unescaped `&`.

Comment (alanfung) [inline tenable_io/helpers/workbench.py]: Going to raise TenableIOException when there is a parser error.

File: tenable_io/api/workbenches.py (hunk 4 above)
Comment (alanfung) [inline tenable_io/api/workbenches.py]: This should already be done in the helper.

File: tests/integration/api/test_workbenches.py (hunk 6 above)
Comment (alanfung) [inline tests/integration/api/test_workbenches.py]: Seems like it might be an issue with the request library. I'll remove `None`.

**Pull request merged**